python/pyspark/sql/session.py

```diff
@@ -373,16 +375,16 @@ def _createFromRDD(self, rdd, schema, samplingRatio):
         rdd = rdd.map(schema.toInternal)
         return rdd, schema
 
-    def _createFromLocal(self, data, schema):
+    def _createFromLocal(self, prepareFunc, data, schema):
         """
         Create an RDD for DataFrame from a list or pandas.DataFrame, returns
         the RDD and schema.
         """
-        # make sure data could consumed multiple times
-        if not isinstance(data, list):
-            data = list(data)
 
         if schema is None or isinstance(schema, (list, tuple)):
+            # make sure data could consumed multiple times
+            if not isinstance(data, list):
```

Comment:
here we can convert to list because there will be several iterations on the data